node/pkg/tss/implementation.go

```diff
@@ -0,0 +1,381 @@
+// If the storage file hadn't contained symetric keys, it'll compute them.
+func NewGuardianStorageFromFile(storagePath string) (*GuardianStorage, error) {
+	var storage GuardianStorage
+	if err := storage.load(storagePath); err != nil {
+		return nil, err
+	}
+
+	return &storage, nil
+}
+
+// ProducedSignature lets a listener receive the output signatures once they're ready.
+func (t *Engine) ProducedSignature() <-chan *common.SignatureData {
+	return t.fpSigOutChan
+}
+
+// ProducedOutputMessages ensures a listener can send the output messages to the network.
+func (t *Engine) ProducedOutputMessages() <-chan *gossipv1.GossipMessage {
+	return t.gossipOutChan
+}
+
+// BeginAsyncThresholdSigningProtocol used to start the TSS protocol over a specific msg.
+func (t *Engine) BeginAsyncThresholdSigningProtocol(vaaDigest []byte) error {
+	if t == nil {
+		return fmt.Errorf("tss engine is nil")
+	}
+	if t.started.Load() != started {
+		return fmt.Errorf("tss engine hasn't started")
+	}
+
+	if t.fp == nil {
+		return fmt.Errorf("tss engine is not set up correctly, use NewReliableTSS to create a new engine")
+	}
+
+	if len(vaaDigest) != digestSize {
+		return fmt.Errorf("vaaDigest length is not 32 bytes")
+	}
+
+	d := party.Digest{}
+	copy(d[:], vaaDigest)
+
+	// fmt.Printf("guardian %v started signing protocol: %v\n", t.GuardianStorage.Self.Index, d)
+	return t.fp.AsyncRequestNewSignature(d)
+}
+
+// TODO: get a signature output channel, so the guardian can listen to outputs from this tssEngine.
+func NewReliableTSS(storage *GuardianStorage) (ReliableTSS, error) {
+	if storage == nil {
+		return nil, fmt.Errorf("the guardian's tss storage is nil")
+	}
+	// fmt.Println("guardian storage loaded, threshold is:	", storage.Threshold)
+
+	fpParams := party.Parameters{
+		SavedSecrets: storage.SavedSecretParameters,
+		PartyIDs:     storage.Guardians,
+		Self:         storage.Self,
+		Threshold:    storage.Threshold,
+		WorkDir:      "",
+		MaxSignerTTL: time.Minute * 5,
+	}
+
+	fp, err := party.NewFullParty(&fpParams)
+	if err != nil {
+		return nil, err
+	}
+
+	fpOutChan := make(chan tss.Message) // this one must listen on it, and output to the p2p network.
+	fpSigOutChan := make(chan *common.SignatureData)
+	fpErrChannel := make(chan *tss.Error)
+
+	t := &Engine{
+		GuardianStorage: *storage,
+
+		fp:           fp,
+		fpOutChan:    fpOutChan,
+		fpSigOutChan: fpSigOutChan,
+		fpErrChannel: fpErrChannel,
+
+		gossipOutChan: make(chan *gossipv1.GossipMessage),
+		started:       atomic.Uint32{}, // default value is 0
+	}
+
+	return t, nil
+}
+
+// Start starts the TSS engine, and listens for the outputs of the full party.
+func (t *Engine) Start(ctx context.Context) error {
+	if t == nil {
+		return fmt.Errorf("tss engine is nil")
+	}
+
+	if !t.started.CompareAndSwap(notStarted, started) {
+		return fmt.Errorf("tss engine has already started")
+	}
+
+	t.ctx = ctx
+
+	if err := t.fp.Start(t.fpOutChan, t.fpSigOutChan, t.fpErrChannel); err != nil {
```

Comment:
there's a flow here where t.started is set to true but t.fp fails to start and you return an error. 
t.started is still true, so starting this engine will never happen again.
I think this is okay though, but worth double checking.

Reply:
I thought it was okay to keep it as `started` because it returned an error, 
But you are right; it's probably safer to change it back to `notStarted`.